enhancements/ingress/aws-load-balancer-operator.md

```diff
@@ -0,0 +1,394 @@
+Load Balancers. The out-of-tree controller supports integration of Kubernetes
+resources with more of the AWS services like WAFs and the DDoS prevent service
+[Shield](https://docs.aws.amazon.com/waf/latest/developerguide/ddos-overview.html).
+It also enables traffic routing from the load balancer directly to the pods of
+an Ingress or Service if the cluster uses the [AWS VPC Container Networking
+Interface](https://github.com/aws/amazon-vpc-cni-k8s). In addition it also
+introduces 2 Custom Resources namely
+[`IngressClassParams`](https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.3/guide/ingress/ingress_class/#ingressclassparams)
+and
+[`TargetGroupBindings`](https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.3/guide/targetgroupbinding/targetgroupbinding/).
+_IngressClassParams_ is used for specifying custom parameters for Ingresses of a
+certain class.  The controller uses _TargetGroupBinding_ internally and it can
+also be used by users to manage the load balancer externally. The
+aws-load-balancer will henceforth be referred to as **“lb-controller”** for the
+sake of brevity.
+
+## Motivation
+
+Implement an operator, tentatively called `aws-load-balancer-operator`, which
+deploys and manages an instance of the lb-operator. The operator will be
+distributed through Operator Hub.
+
+### Goals
+* Implement operator which can be installed and used in OCP/OKD.
+
+### Non-Goals
+* Implementing support for ALB for Openshift Routes
+* Support for multiple installations of the controller i.e. multiple custom
+  resources.
+* Support for `ip` mode for Ingress resources
+* Support for NLB backed Services through the lb-controller.
+
+## Proposal
+
+### User Stories
+
+#### TLS Termination and DDoS protection
+
+The user wants to route traffic for the domain example.com to pods of Service
+example-service and have TLS termination and DDoS protection on the load
+balancer. The user deploys the aws-alb-operator and creates an instance of the
+resource AWSLoadBalancerController as follows:
+
+```yaml
+kind: AWSLoadBalancerController
+group: networking.openshift.io/v1beta1
+metadata:
+  name: cluster
+spec:
+  autoTagSubnet: true
+  ingressClass: tls-termination
+```
+Then they create an ingress resource with the following schema:
+```yaml
+apiVersion: networking.k8s.io/v1
+kind: Ingress
+metadata:
+  name: example
+  annotations:
+    alb.ingress.kubernetes.io/scheme: internet-facing
+    alb.ingress.kubernetes.io/certificate-arn: arn:aws:acm:us-west-2:xxxxx
+    alb.ingress.kubernetes.io/shield-advanced-protection: true
+spec:
+  ingressClassName: tls-termination
+  rules:
+  - host: example.com
+    http:
+        paths:
+          - path: /
+            pathType: Exact
+            backend:
+              service:
+                name: example-service
+                port:
+                  number: 80
+```
+The controller then creates an application load balancer with an HTTPS listener
```

Comment:
The reason it creates an HTTPS listener only (and not also an HTTP listener) is because `alb.ingress.kubernetes.io/certificate-arn` is specified and `alb.ingress.kubernetes.io/listen-ports` is omitted—correct?  It would be helpful to note this for people who are unfamiliar with aws-load-balancer-controller.